x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepositoryPlugin.java

```diff
@@ -6,25 +6,153 @@
+                final String delegateType = DELEGATE_TYPE_SETTING.get(metadata.settings());
+                if (Strings.hasLength(delegateType) == false) {
+                    throw new IllegalArgumentException("Repository setting [" + DELEGATE_TYPE_SETTING.getKey() + "] must be set");
+                }
+                if (REPOSITORY_TYPE_NAME.equals(delegateType)) {
+                    throw new IllegalArgumentException(
+                        "Cannot encrypt an already encrypted repository. ["
+                            + DELEGATE_TYPE_SETTING.getKey()
+                            + "] must not be equal to ["
+                            + REPOSITORY_TYPE_NAME
+                            + "]"
+                    );
+                }
+                final Repository.Factory factory = typeLookup.apply(delegateType);
+                if (null == factory || false == SUPPORTED_ENCRYPTED_TYPE_NAMES.contains(delegateType)) {
+                    throw new IllegalArgumentException("Unsupported delegate repository type [" + DELEGATE_TYPE_SETTING.getKey() + "]");
+                }
+                final String repositoryPasswordName = PASSWORD_NAME_SETTING.get(metadata.settings());
+                if (Strings.hasLength(repositoryPasswordName) == false) {
+                    throw new IllegalArgumentException("Repository setting [" + PASSWORD_NAME_SETTING.getKey() + "] must be set");
+                }
+                final char[] repositoryPassword = repositoryPasswordsMap.get(repositoryPasswordName);
+                if (repositoryPassword == null) {
+                    throw new IllegalArgumentException(
+                        "Secure setting ["
+                            + ENCRYPTION_PASSWORD_SETTING.getConcreteSettingForNamespace(repositoryPasswordName).getKey()
+                            + "] must be set"
+                    );
+                }
+                final Repository delegatedRepository = factory.create(
+                    new RepositoryMetadata(metadata.name(), delegateType, metadata.settings())
+                );
+                if (false == (delegatedRepository instanceof BlobStoreRepository) || delegatedRepository instanceof EncryptedRepository) {
+                    throw new IllegalArgumentException("Unsupported delegate repository type [" + DELEGATE_TYPE_SETTING.getKey() + "]");
+                }
+                if (false == getLicenseState().isEncryptedSnapshotAllowed()) {
+                    logger.warn(
+                        "Encrypted snapshots are not allowed for the currently installed license."
+                            + "Snapshots to the ["
+                            + metadata.name()
+                            + "] encrypted repository are not permitted."
+                            + "All the other operations, including restore, work without restrictions.",
+                        LicenseUtils.newComplianceException("encrypted snapshots")
```

Comment:
A few small suggestions:

- I think printing the license mode will assist understanding the message.
- It needs whitespace between the joined sentences
- Our typical idiom is to use parameters rather than string concatenation.

```suggestion
                    logger.warn(new ParameterizedMessage(
                        "Encrypted snapshots are not allowed for the currently installed license [{}]."
                            + " Snapshots to the [{}] encrypted repository are not permitted."
                            + " All the other operations, including restore, work without restrictions.",
                            getLicenseState().getOperationMode().description(), metadata.name()),
                        LicenseUtils.newComplianceException("encrypted snapshots")
```

Reply:
Thank you.